### ISSUE TITLE
Records deleted when new records are assigned to belonging association

### ISSUE THREAD

**Issue by delonnewman** [rails/rails#41261]
### Steps to reproduce
First, I should say, I'm not 100% certain that the maintainers would consider this a bug, but it does seem to be unexpected (at least for me) and undocumented behavior (from what I can see in the [documentation](https://api.rubyonrails.org/v5.2.4.4/classes/ActiveRecord/Associations/ClassMethods.html#method-i-has_many)).

```ruby
class Request < ApplicationRecord
  has_many :items, class: 'RequestItem', dependent: :delete_all
end

class RequestItem < ApplicationRecord
  belongs_to :request
end

Request.new(id: 1, items: [RequestItem.new(start_at: Time.now, end_at: Time.now + 36000)])
# also
r = Request.new(id: 1)
r.items = [RequestItem.new(start_at: Time.now, end_at: Time.now + 36000)]
```
Where a request with ID 1 exists in the database, any existing request items are deleted.

I'm using activerecord 5.2.4.4, and activerecord-sqlserver-adapter 5.2.0
I've also replicated the behavior with sqlite here: https://github.com/delonnewman/destroy_all/blob/master/test/models/request_test.rb

Also, if this is a bug, I'd be willing to work towards a patch.
Thank you so much for your work!

### Expected behavior
It seems unexpected for records to be deleted without explicitly deleting the parent record. If deleting the records in this case is actually the desired behavior one might expect that it would at least persist the new records that have been assigned (it does not).

### Actual behavior
The "belonging" records are all deleted. After the assignment the "has_many" association collection is empty.

### System configuration
**Rails version**: 5.2.4.4

**Ruby version**: 2.7.2

**Comment by intrip** [comment]
I don't think this is a bug. This happens because when you assign directly an `Array` to a `has_many` association [you are meant to replace all the related existing records](https://guides.rubyonrails.org/v5.2/association_basics.html#methods-added-by-has-many-collection-objects).

**Comment by delonnewman** [comment]
Thank you for the reply.

I share your reluctance to consider it a bug, there's one problem with that--the records are not replaced they're deleted (i.e. there are no records in the association after the assignment). I included a link to a repo that reproduces the behavior in the original message.

Thanks again.

**Comment by intrip** [comment]
This doesn't look to be a bug:
- The problem is that you are not saving the record `r` in the end: when you do `r.items = [RequestItem.new(...)]` the old items are deleted but the new items aren't saved unless you call `r.save`.
- Also: since you are creating the record with `Request.new` `r.save` will raise a unique primary key error, you should use `r = Request.find(1)` in order to make it work as expected.
